### ISSUE TITLE
A test fails on ruby/ruby master when the path of TMPDIR is long

### ISSUE THREAD

**Issue by mame** [rubygems/rubygems#3737]
Hi @deivid-rodriguez 

`test/rubygems/test_gem_package.rb` fails on ruby/ruby master when the path of TMPDIR is long.

https://rubyci.org/logs/rubyci.s3.amazonaws.com/solaris11-gcc/ruby-master/log/20200617T130007Z.fail.html.gz
```
  1) Failure:
TestGemPackage#test_extract_symlink_parent_doesnt_delete_user_dir [/export/home/chkbuild/chkbuild-gcc/tmp/build/20200617T130007Z/ruby/test/rubygems/test_gem_package.rb:620]:
--- expected
+++ actual
@@ -1 +1 @@
-"installing into parent path /export/home/chkbuild/chkbuild-gcc/tmp/build/20200617T130007Z/ruby/tmp/test_rubygems_15916/extract/user/dir of /export/home/chkbuild/chkbuild-gcc/tmp/build/20200617T130007Z/ruby/tmp/test_rubygems_15916/extract/subdir is not allowed"
+"installing into parent path link/dir of /export/home/chkbuild/chkbuild-gcc/tmp/build/20200617T130007Z/ruby/tmp/test_rubygems_15916/extract/subdir is not allowed"
```

`/export/home/chkbuild/chkbuild-gcc/tmp/build/20200617T130007Z/ruby/tmp` is the path of TMPDIR on this environment.

The test `test_extract_symlink_parent_doesnt_delete_user_dir` creates a temporal directory in TMPDIR and attempts to add a symlink to the directory into a tarball, but tar format limits the max length of path up to 99, so the path is truncated, which makes the path check of `Gem::Package#install_location` pass.  Eventually the execution is stopped by an error, but the error message is slightly different (absolute path and relative one) from the expectation of the test.

The issue is reproducible in rubygems/rubygems:

```
$ mkdir /tmp/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
$ TMPDIR=/tmp/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/ ruby test/rubygems/test_gem_package.rb

...

  2) Failure:
TestGemPackage#test_extract_symlink_parent_doesnt_delete_user_dir [test/rubygems/test_gem_package.rb:620]:
--- expected
+++ actual
@@ -1 +1 @@
-"installing into parent path /tmp/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/test_rubygems_63202/extract/user/dir of /tmp/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/test_rubygems_63202/extract/subdir is not allowed"
+"installing into parent path link/dir of /tmp/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/test_rubygems_63202/extract/subdir is not allowed"

...
```

I tentatively added into ruby/ruby a skip statement when the path is very long (> 90).  https://github.com/ruby/ruby/commit/e30ff63fd4eb24fd638259cb10638668f0f2ceeb

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by deivid-rodriguez** [comment]
Thanks @mame!

I think it's fine to skip the test like you did for now. Can you open a PR?

**Comment by nobu** [comment]
When the path is short enough, the exception is raised from `install_location`, but when it is too long, raised from `mkdir_p_safe`.
I'm afraid that this may not be intended too.

**Comment by deivid-rodriguez** [comment]
I can also try to investigate it, but I'd like to be in sync with ruby-core anyways :)

**Comment by deivid-rodriguez** [comment]
I tried to reproduce this with your steps but I failed... I get the same results with or without passing `TMPDIR`, which makes sense to me since our test suite sets `TMPDIR` at the beginning of each test.

I think maybe to reproduce I need to move my rubygems working copy to a very nested folder?

**Comment by nobu** [comment]
`TMPDIR` doesn't affect this test, as it uses `tmp` directory always.
I reproduced this issue by making `tmp` a very long symlink.

**Comment by deivid-rodriguez** [comment]
Yes, that made it reproduce, thanks! :+1:

**Comment by nobu** [comment]
Where `destination_user_dir = File.join @destination, 'user'`, this path name can be too long, then `tar.add_symlink 'link', destination_user_dir, 16877` may make truncated `linkname` entry.
At extracting, `File.realpath` in `Gem::Package#install_location` fails with `ENOENT` due to the truncation, and `install_location` wrongly succeeds as "it's fine if the destination doesn't exist".

**Comment by nobu** [comment]
Although I'm not sure if this test has to use an absolute link name, how about this?

```diff
diff --git a/test/rubygems/test_gem_package.rb b/test/rubygems/test_gem_package.rb
index 1a0482cb1e..ff6bca56fc 100644
--- a/test/rubygems/test_gem_package.rb
+++ b/test/rubygems/test_gem_package.rb
@@ -606,7 +606,7 @@
     FileUtils.mkdir_p destination_user_subdir
 
     tgz_io = util_tar_gz do |tar|
-      tar.add_symlink 'link', destination_user_dir, 16877
+      tar.add_symlink 'link', '../user', 0755
       tar.add_symlink 'link/dir', '.', 16877
     end
 
```

**Comment by deivid-rodriguez** [comment]
The problem is that this test was added to prevent a security issue where someone would maliciously create a package with a symlink pointing to an absolute well-know system file, and rubygems would remove it before extracting the package. See https://blog.rubygems.org/2019/03/05/security-advisories-2019-03.html, and e853acb604ffce03622a0aa2a528d1f7154f055e. So, I think we need an absolute path there to properly protect ourselves from that issue?

**Comment by deivid-rodriguez** [comment]
Well, I guess you could do "path guessing" with a relative target too, so maybe they are equivalent :man_shrugging:.

**Comment by nobu** [comment]
Then the scenario is like the following?

```ruby
    if destination_user_dir.size > 100
      destination_user_dir = '/usr'
      destination_user_subdir = '/usr/bin'
      if File.writable?(destination_user_dir) or !File.directory?(destination_user_subdir)
        skip "TMPDIR seems too long to add it as symlink into tar"
      end
    end

    tgz_io = util_tar_gz do |tar|
      tar.add_symlink 'link', destination_user_dir, 16877
      tar.add_symlink 'link/'+File.basename(destination_user_subdir), '.', 16877
    end
```

Of course this is risky to run as a test.

**Comment by deivid-rodriguez** [comment]
Yeah, something like that would be the way and conditions to exploit it when the vulnerability was present. In any case, #3738 seems fine to me?

**Comment by deivid-rodriguez** [comment]
I merged #3788 to get in sync, so closing this ticket. I'll try to follow up on improving symlink support with #2836.
